Official Ably Client Broadcaster Doc (#1)
ably-forks/laravel-docs: broadcasting.md

<a name="client-ably"></a>
### Ably

[Laravel Echo](https://github.com/laravel/echo) is a JavaScript library that makes it painless to subscribe to channels and listen for events broadcast by your server-side broadcasting driver.

> Install **official** `ably` package along with `laravel-echo` using NPM. 

Reviewer: I think we should make this section similar to the Pusher channels section:

```suggestion
[Laravel Echo](https://github.com/laravel/echo) is a JavaScript library that makes it painless to subscribe to channels and listen for events broadcast by your server-side broadcasting driver. You may install Echo via the NPM package manager. In this example, we will also install the ably package since we will be using the Ably broadcaster:
```

```shell
npm install --save-dev laravel-echo ably
`````

Once Echo is installed, you are ready to create a fresh Echo instance in your applications JavaScript. A great place to do this is at the bottom of the `resources/js/bootstrap.js` file that is included with the Laravel framework. By default, an example Echo configuration is already included in this file; however, the default configuration in the bootstrap.js file is intended for Pusher. You may copy the configuration below to transition your configuration to Ably

```js
import Echo from 'laravel-echo';
import * as Ably from 'ably';

window.Ably = Ably;
window.Echo = new Echo({
    broadcaster: 'ably',
});

window.Echo.connector.ably.connection.on(stateChange => {
    if (stateChange.current === 'connected') {
        console.log('connected to ably server');
    }
});
```

> You can set custom [clientOptions](https://docs.ably.io/client-lib-development-guide/features/#options) while creating `Echo` instance.

Reviewer: ```suggestion
You can set custom [clientOptions](https://ably.com/docs/api/realtime-sdk#client-options) when creating an `Echo` instance.
```

Reviewer: This comment still stands - we should link to the API reference docs rather than the features spec

Author: Actually, I feel the features spec is misleading. You can take a look at options, I think they are following `ruby` conventions for `snake_case` options.

Author: Updated!

```
    broadcaster: 'ably',
    authEndpoint: 'http://www.localhost:8000/broadcasting/auth'

>     // Additional ably specific options - https://docs.ably.io/client-lib-development-guide/features/#options  

Reviewer: ```suggestion
    // Additional ably specific options - https://ably.com/docs/api/realtime-sdk#client-options  
```

    realtimeHost: 'realtime.ably.com',
    restHost: 'rest.ably.com',
    port: '80',
    echoMessages: true // By default self-echo for published message is false
```
Once you have uncommented and adjusted the Echo configuration according to your needs, you may compile your application's assets:

```shell
npm run dev
```

> > {tip} To learn more about compiling your application's JavaScript assets, please consult the documentation on [Laravel Mix](/docs/{{version}}/mix).

Reviewer: Is this change needed? Doesn't seem to be related to the changes we're making.

Author: Ohh yes, you are right!

### Additional supported features (official ably client)

> **1. Update token expiry (Default 3600, 1 hr)**

Reviewer: ```suggestion
**1. Update token expiry. Default: 3600 seconds (1 hr)**
```

- Update`ABLY_TOKEN_EXPIRY` in `.env` file. 

> - Update `ably` section under `config/broaadcasting.php` with `'token_expiry' => env('ABLY_TOKEN_EXPIRY', 3600)`

Reviewer 2: Typo, `config/broadcasting.php`

Author: Updated!

**2. Modify channel capability**
- Channel access can be changed as per [Channel Capabilities](https://ably.com/docs/core-features/authentication#capability-operations)
```php
  // file - routes/channels.php

>   // for private channel (array is returned as a truthy value)

Reviewer: The sentence "array is returned as a truthy value" is a bit confusing to me, I'm not sure what you're trying to say here.

Author: It has to do with returning either `true` or `false`. When `truthy` value is returned, channel access is given. So, it doesn't matter if the returned value is of type `array` or `object` or `boolean`, all of them are treated as `true` values. It's the same concept of `Truthy` and `Falsy` values in js -> https://www.sitepoint.com/javascript-truthy-falsy/

Reviewer: OK yeah I get that but I still think this could be clearer - maybe it should say:
```suggestion
  // for private channel (If a falsy value is returned then channel access is not permitted for the client)
```

also FWIW empty arrays are falsy in PHP

Author: Updated to ->
`// for private channel (Access is allowed for truthy values and denied for falsy values)`

>
>Once Echo is installed, you are ready to create a fresh Echo instance in your application's JavaScript. A great place to do this is at the bottom of the `resources/js/bootstrap.js` file that is included with the Laravel framework. By default, an example Echo configuration is already included in this file; however, the default configuration in the `bootstrap.js` file is intended for Pusher. You may copy the configuration below to transition your configuration to Ably:
>
>```js
>import Echo from 'laravel-echo';
>
>window.Pusher = require('pusher-js');
>
>window.Echo = new Echo({
>    broadcaster: 'pusher',
>    key: process.env.MIX_ABLY_PUBLIC_KEY,
>    wsHost: 'realtime-pusher.ably.io',
>    wsPort: 443,
>    disableStats: true,
>    encrypted: true,
>});
>```
>
>Note that our Ably Echo configuration references a `MIX_ABLY_PUBLIC_KEY` environment variable. This variable's value should be your Ably public key. Your public key is the portion of your Ably key that occurs before the `:` character.
>

> >Update php-Laravel config. - Set `ABLY_PUSHER_ADAPTER` as **true** in `.env` file. Also, make sure `ably` section under `config/broadcasting.php` is updated with `'pusher_adapter' => env('ABLY_PUSHER_ADAPTER', true)`

Reviewer 2: Second part is not required, and isn't common practice in Laravel, as it's the default value and should be left at `false`. The purpose of the `.env` file and `env` function is to allow users to use custom values instead of the default specified in `config/broadcasting.php` or another config file.

Author: Updated!

### Migrating from deprecated ably-pusher adapter to ably

> - Latest Ably broadcaster is fully compatible with old Ably-Pusher broadcaster.

Reviewer: ```suggestion
- The latest Ably broadcaster is fully compatible with the deprecated Ably-Pusher broadcaster.
```

> - Only **Leaving the channel**, [Use Ably Channel Namespace](https://ably.com/docs/general/channel-rules-namespaces) conventions

Reviewer: ```suggestion
- The only API difference is for **leaving a channel**, you must [use Ably Channel Namespace](https://ably.com/docs/general/channel-rules-namespaces) conventions
```
